### PR TITLE
Revert "Revert "Sdit 2182 create duplicate logging from the red index updates""

### PR DESCRIPTION
**Pull request by steverendell**
Reverts ministryofjustice/hmpps-prisoner-search#460 which reverted #453

### CODE REVIEW

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt [hunk 1]
```diff
@@ -107,21 +152,19 @@ class PrisonerSynchroniserService(
...
+              if (updated) {
+                TelemetryEvents.INCENTIVE_UPDATED
+              } else {
+                TelemetryEvents.INCENTIVE_OPENSEARCH_NO_CHANGE
+              },
+              prisonerNumber = prisonerNo,
+              bookingId = bookingId,
+              eventType = eventType,
+            )
+          }
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt:167]: Used also {} as suggested in prev PR

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt [hunk 2]
```diff
@@ -70,18 +80,35 @@ class PrisonerSynchroniserService(
...
           TelemetryEvents.RED_PRISONER_UPDATED,
           prisonerNumber = ob.offenderNo,
           bookingId = ob.bookingId,
           eventType = eventType,
         )
         // If the current booking id changed, some domain data needs updating too, retrieving using the new booking id
-        if (summary.prisoner?.bookingId?.toLong() != ob.bookingId) {
+        if (summary.prisoner.bookingId?.toLong() != ob.bookingId) {
           reindexIncentive(ob.offenderNo, SyncIndex.RED, eventType)
+          reindexRestrictedPatient(ob.offenderNo, ob, SyncIndex.RED, eventType)
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt:108]: RP update was missing

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt [hunk 3]
```diff
@@ -93,7 +120,25 @@ class PrisonerSynchroniserService(
...
+      alertsUpdatedEventService.generateAnyEvents(null, prisoner, red = true)
+
+      // simulating domainEventEmitter.emitPrisonerCreatedEvent(ob.offenderNo)
+      telemetryClient.trackPrisonerEvent(
+        TelemetryEvents.RED_SIMULATE_PRISONER_CREATED_EVENT,
+        prisonerNumber = ob.offenderNo,
+        bookingId = ob.bookingId,
+        eventType = eventType,
+      )
+
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt:141]: Cannot use reindex() as that has real event generation included. This code does the same job but with red-specific events and logging

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt [hunk 4]
```diff
@@ -187,7 +228,13 @@ class PrisonerSynchroniserService(
 
       indices.map { prisonerRepository.save(prisoner, it) }
       if (label == PrisonerDifferencesLabel.GREEN_BLUE) {
-        prisonerDifferenceService.handleDifferences(existingPrisoner, ob, prisoner, "REFRESH")
+        if (prisonerDifferenceService.handleDifferences(existingPrisoner, ob, prisoner, "REFRESH")) {
+          prisonerMovementsEventService.generateAnyEvents(existingPrisoner, prisoner, ob)
+          alertsUpdatedEventService.generateAnyEvents(existingPrisoner, prisoner)
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerSynchroniserService.kt:233]: generateAnyEvents() calls moved up

File: hmpps-prisoner-search-indexer/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/events/HmppsDomainEventsEmitterIntTest.kt [hunk 5]
```diff
@@ -54,25 +54,25 @@ import uk.gov.justice.hmpps.sqs.countAllMessagesOnQueue
 import java.time.Duration
 
 class HmppsDomainEventsEmitterIntTest : IntegrationTestBase() {
-  @SpyBean
+  @MockitoSpyBean
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/events/HmppsDomainEventsEmitterIntTest.kt:57]: SpyBean now deprecated!

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**